torch/distributed/distributed_c10d.py

```diff
@@ -1598,6 +1619,12 @@ def destroy_process_group(group: Optional[ProcessGroup] = None):
         pg._wait_for_pending_works()
 
     if group is None or group == GroupMember.WORLD:
+        if _abort_in_destroy_pg():
+            # shutdown all backends in the order of pg names. shutting down in order because
+            # ncclCommAbort() was a 'collective' call in some versions of NCCL.
+            for pg_to_shutdown in sorted(_world.pg_names, key=lambda x: _world.pg_names[x], reverse=True):
```

Comment:
I am not sure if we are completely deadlock free OK to rely on PG creation order, since nccl communicator may be created in different order, i.e., created when the first collective call is made with that pg.
```
pg0=create_group()
pg1=create_group()
...
allreduce(pg1)
// created pg1->comm
allreduce(pg0)
// created pg0->comm
...
```
Maybe better to based on actual nccl communicator creation, to be safe?

Reply:
good point, i forgot about this!

Reply:
Good point. but want to point out that this is 'best effort' ordering, it is just better than random order as of existing destruction order. it is not easy for python (e.g., without a c++ class level map) to track the actual creation order of ncclComms. 

Also Does it really matter to follow the exact reverse order of creation? I think as long as all ranks are terminating the PGs in the same order, we should be fine.

Reply:
my theory about ordering (which i totally made up) was that a child group might need to be destroyed before the parent group it came from.  I don't know if order matters besides that.  (if you have 3 child groups from one parent, does it matter which order you destroy the children in? 

Agree that getting consistent ordering across ranks is a basic requirement and an improvement.

Reply:
I considered this as resolved then? from unit tests of 2.19, the exact order within the same rank does not make a difference, but we can keep an eye on this, if it really matter, we can always address it later

Reply:
Also, we need to consider if the children's comm is 'split' from the parent comm.  if not split, the destruction order might not matter. If children is split from parent, the real coms creation order must still be parent first (PG0), children later (e.g. PG1), which is the same as the python layer creation order, irrespective of which PG calls the collective first.  This is a requirement of ncclCommSplit (no outstanding op on parent comms: https://docs.nvidia.com/deeplearning/nccl/user-guide/docs/api/comms.html#ncclcommsplit)
https://github.com/pytorch/pytorch/blob/fa071a2e1b6617ba722d639747f1d885017b0e86/torch/csrc/distributed/c10d/NCCLUtils.hpp#L299

Reply:
I agree that we are good to destroy communicator in arbitrary order, as long as every rank follows the same order. I was thinking some potential circular deadlock case, where ranks create partially overlapped communicator in different order (e.g., FSDP PG + TP PG). But I guess it would still be fine, since we already force every rank to follow the same destroy order.